internal/util/recoverable.go

```diff
@@ -0,0 +1,113 @@
+package util
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"runtime/debug"
+	"sync"
+	"time"
+)
+
+var (
+	ErrServiceStopped = fmt.Errorf("service stopped")
+)
+
+type Doable interface {
+	Do() error
+	Stop()
+}
+
+func NewRecoverableService(svc Doable, logger *log.Logger) *RecoverableService {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &RecoverableService{
+		service: svc,
+		stopped: make(chan stoppedService, 1),
+		log:     logger,
+		ctx:     ctx,
+		cancel:  cancel,
+	}
+}
+
+type RecoverableService struct {
+	mu      sync.Mutex
+	running bool
+	service Doable
+	stopped chan stoppedService
+	log     *log.Logger
+	ctx     context.Context
+	cancel  context.CancelFunc
+}
+
+func (m *RecoverableService) Start(_ context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.running {
+		return nil
+	}
+
+	go m.serviceStart()
+	m.running = true
+	return nil
+}
+
+func (m *RecoverableService) Stop() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !m.running {
+		return
+	}
+
+	m.service.Stop()
+	m.cancel()
+	m.running = false
+}
+
+func (m *RecoverableService) serviceStart() {
+	for {
+		select {
+		case st := <-m.stopped:
+			// restart the service
+			if st.Err != nil && errors.Is(st.Err, ErrServiceStopped) {
+				<-time.After(10 * time.Second)
```

Comment:
Should the rest period between restarts be configurable?

Reply:
Maybe. It might also work best as an exponential backoff. I really would like to ensure that hard errors are exposed from this type of service failure. Maybe even go as far as either panicing or throwing errors from the observer. That would be a good bit to figure out.